src/box/memtx_tx.c

```diff
@@ -365,23 +608,123 @@ memtx_tx_handle_conflict(struct txn *breaker, struct txn *victim)
 	}
 }
 
+/**
+ * Calculate size of story with its links.
+ */
+static inline size_t
+memtx_story_size(struct memtx_story *story)
+{
+	struct mempool *pool = &txm.memtx_tx_story_pool[story->index_count];
+	return pool->objsize;
+}
+
+/**
+ * Notify memory manager that a tuple referenced by @a story
+ * was replaced from primary key and that is why @a story
+ * is the only reason why the tuple cannot be deleted.
+ */
+static inline void
+memtx_tx_story_tuple_retain(struct memtx_story *story)
+{
+	assert(!story->tuple_is_retained);
+	assert(story->status < MEMTX_TX_STORY_MAX);
+
+	story->tuple_is_retained = true;
+	struct memtx_tx_stats *stats = &txm.retained_tuple_stats[story->status];
+	size_t tuplesize = tuple_size(story->tuple);
+	memtx_tx_stats_collect(stats, tuplesize);
+}
+
+/**
+ * Notify memory manager that a tuple referenced by @a story
+ * was placed to primary key.
+ */
+static inline void
+memtx_tx_story_tuple_release(struct memtx_story *story)
```

Comment:
I think the names are misleading: I'd rather it be something like
```suggestion
memtx_tx_story_untrack_retained_tuple(struct memtx_story *story)
```

Reply:
agree, fixed